### ISSUE TITLE
CXDirection pass should not generate gates outside basis_gates

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#3972]
@kdk [noticed that](https://github.com/Qiskit/qiskit-terra/issues/3970#issuecomment-598826238) `transpile` might return a circuit that is not in the given basis:
> [...] `CXDirection` is inserting Hadamards, which can't be unrolled to U1 gates (only U2 or U3) [...]
> So there is an issue here in that if the circuit can't be expressed in terms of the requested basis, we should raise an error instead of returning a circuit in the wrong basis.

Here an example:
```
circuit = QuantumCircuit(3)
circuit.cx(0, 2)

basis = ['u1', 'cx'].  # notice that u2 is not in the basis.
coupling_map = CouplingMap([(0, 1), (1, 2)])
result = transpile(circuit,
                   optimization_level=1,
                   basis_gates=basis,
                   coupling_map=coupling_map,
                   seed_transpiler=42)
print(result)
```
```
            ┌──────────┐                                                   »
q_2 -> 0 |0>┤ U2(0,pi) ├────────────────────────────────────────────────■──»
            └──────────┘┌──────────┐     ┌──────────┐     ┌──────────┐┌─┴─┐»
q_1 -> 1 |0>─────■──────┤ U2(0,pi) ├──■──┤ U2(0,pi) ├──■──┤ U2(0,pi) ├┤ X ├»
               ┌─┴─┐    ├──────────┤┌─┴─┐├──────────┤┌─┴─┐└──────────┘└───┘»
q_0 -> 2 |0>───┤ X ├────┤ U2(0,pi) ├┤ X ├┤ U2(0,pi) ├┤ X ├─────────────────»
               └───┘    └──────────┘└───┘└──────────┘└───┘                 »
```

**Comment by mtreinish** [comment]
I wonder if this is something we could fit in under https://github.com/Qiskit/qiskit-terra/pull/3604

**Comment by 1ucian0** [comment]
I think so...

**Comment by kdk** [comment]
Following up on some offline discussion, there is still an issue here (though the original example currently raises from `Optimize1qGates` at `optimization_level`s other than zero, and both the BasisTranslator and Unroller should raise if asked to translate to a basis they can't reach).

The remaining issue is that `CXDirection` runs _after_ unrolling/translation and inserts `U2Gate`s instead of Hadamards. 

![image](https://user-images.githubusercontent.com/2241698/98877547-bd7bfa80-244e-11eb-8c9b-f8096539d0ce.png)

Probably the right fix is to move `CXDirection` to run after the routing passes but before the translation passes.

**Comment by kdk** [comment]
Will be resolved as part of #5609 .
